documentation/boltspec_reference.md

```diff
@@ -0,0 +1,1117 @@
+  hash of key-value pairs where each key is a target and the value is the value
+  for that target's `Result`.
+
+  ```ruby
+  allow_task('pet_dog').return_for_targets(
+    'target1' => { 'happy' => true },
+    'target2' => { 'happy' => false }
+  )
+  ```
+
+- `return(&block) { |targets, task, params| ... }`
+
+  Invokes a block to construct the `ResultSet` returned by the `run_task`
+  function.
+
+  ```ruby
+  allow_task('pet_dog').return do |targets, task, params|
+    results = targets.map do |target|
+      Bolt::Result.new(target, value: { 'happy' => true })
+    end
+
+    Bolt::ResultSet.new(results)
+  end
+  ```
+
+- `error_with(error)`
+
+  Sets the error hash for each target's `Result` when the task is run. Returns a
+  `Bolt::ResultSet` object.
+
+  ```ruby
+  allow_task('pet_dog').error_with('msg' => 'There are no German Shepherds to pet.')
+  ```
+
+### `allow_upload`
+
+The `allow_upload` function stubs the [`upload_file`
+function](plan_functions.md#upload-file). It accepts a single parameter: the
+path to a local file to upload.
+
+```ruby
+allow_upload('sshd_config')
+```
+
+The `allow_upload` function accepts the following modifiers:
+
+- `be_called_times(number)`
+
+  The test fails if the file is uploaded more than _number_ times.
+
+  ```ruby
+  allow_upload('sshd_config').be_called_times(1)
+  ```
+
+- `not_be_called`
+
+  The test fails if the file is uploaded.
+
+  ```ruby
+  allow_upload('sshd_config').not_be_called
+  ```
+
+- `with_targets(targets)`
+
+  The target or list of targets that the file can be uploaded to. The test
+  fails if the file is uploaded to a different list of targets.
+
+  ```ruby
+  allow_upload('sshd_config').with_targets(['target1', 'target2', 'target3'])
+  ```
+
+- `with_params(parameters)`
+
+  The [options](plan_functions.md#upload-file) that can be passed to the
+  `upload_file` function. The test fails if the file is uploaded with a
+  different set of  options.
+
+  ```ruby
+  allow_upload('sshd_config').with_params({ '_run_as' => 'root' })
+  ```
+
+- `with_destination(destination)`
+
+  The destination path that the file is uploaded to. The test fails if the
+  file is uploaded to a different location.
+
+  ```ruby
+  allow_upload('sshd_config').with_destination('/etc/ssh/sshd_config')
+  ```
+
+- `return(&block) { |targets, source, destination, params| ... }`
+
+  Invokes a block to construct the `ResultSet` returned by the `upload_file`
+  function.
+
+  ```ruby
+  allow_upload('sshd_config').return do |targets, source, destination, params|
+    results = targets.map do |target|
+      Bolt::Result.new(target, value: { 'path' => File.join(destination, source) })
+    end
+
+    Bolt::ResultSet.new(results)
+  end
+  ```
+
+- `error_with(error)`
+
+  Sets the error hash for each target's `Result` when the file is uploaded.
+  Returns a `Bolt::ResultSet` object.
+
+  ```ruby
+  allow_upload('sshd_config').error_with('msg' => 'Not authorized')
+  ```
```

Comment:
Similar to the testing doc, I think it's nice to bookend these with a related info section:

```
📖 Related information

- For more information on using `BoltSpec`, see [Testing plans](testing_plans.md).
```